app/models/screened_ip_address.rb

```diff
@@ -94,55 +106,59 @@ def self.block_admin_login?(user, ip_address)
-       WHERE action_type = #{ScreenedIpAddress.actions[:block]}
-         AND family(ip_address) = 4
-         AND masklen(ip_address) = 32
-    GROUP BY ip_range
-      HAVING COUNT(*) >= :min_count
-    SQL
-  end
+  def self.subnets(family, masklen)
+    base_masklen = case family
+                   # every IPv4 is important and represents a single user
+                   when 4 then 32
+                   # IPv6 are less important and ISPs assign blocks of /64 to a
+                   # single user we set the base masklen to 72 to be more
+                   # similar to IPv4 so we can use the algorithm for rolling up
+                   # 72 (the base masklen) - 64 (the first rollup) for IPv6
+                   # is equal to
+                   # 32 (the base masklen) - 24 (the first rollup) for IPv4
+                   when 6 then 72
+    end
 
-  def self.star_star_subnets_query
-    @star_star_subnets_query ||= <<~SQL
-      WITH weighted_subnets AS (
-        SELECT network(inet(host(ip_address) || '/16'))::text AS ip_range,
-               CASE masklen(ip_address)
-                 WHEN 32 THEN 1
-                 WHEN 24 THEN :roll_up_weight
-                 ELSE 0
-               END AS weight
-          FROM screened_ip_addresses
-         WHERE action_type = #{ScreenedIpAddress.actions[:block]}
-           AND family(ip_address) = 4
+    sql = <<~SQL
+      WITH weighted_ips AS (
+        SELECT ip_address,
+               network(inet(host(ip_address) || '/' || :masklen))::text subnet,
+               greatest(1, pow(2, :base_masklen - masklen(ip_address)) * :weight)::int weight
+        FROM screened_ip_addresses
+        WHERE family(ip_address) = :family AND
+              masklen(ip_address) > :masklen AND
+              action_type = :blocked
       )
-      SELECT ip_range
-        FROM weighted_subnets
-    GROUP BY ip_range
-      HAVING SUM(weight) >= :min_count
+      SELECT subnet
+      FROM weighted_ips
+      GROUP BY subnet
+      HAVING SUM(weight) >= :threshold
     SQL
-  end
-
-  def self.star_subnets
-    min_count = SiteSetting.min_ban_entries_for_roll_up
-    DB.query_single(star_subnets_query, min_count: min_count)
-  end
 
-  def self.star_star_subnets
-    weight = SiteSetting.min_ban_entries_for_roll_up
-    DB.query_single(star_star_subnets_query, min_count: 10, roll_up_weight: weight)
+    weight = SiteSetting.min_ban_entries_for_roll_up.to_f / 100
+
+    DB.query_single(
+      sql,
+      family: family,
+      masklen: masklen,
+      base_masklen: base_masklen,
+      weight: weight,
+      blocked: ScreenedIpAddress.actions[:block],
+      threshold: (2**(base_masklen - masklen) * weight).round
+    )
   end
 
   def self.roll_up(current_user = Discourse.system_user)
-    subnets = [star_subnets, star_star_subnets].flatten
-
-    StaffActionLogger.new(current_user).log_roll_up(subnets) unless subnets.blank?
+    subnets = ROLLED_UP_BLOCKS.map do |family, masklen|
```

Comment:
I think you can do `flat_map` instead of doing `map` here and later doing `flatten(1)`

Reply:
That's clever, thanks!